### DISCUSSION TITLE
Custom agents?

### DISCUSSION THREAD

**Discussion by ssy080** [arabcoders/watchstate#545] [General] [answered]
Hello,

What's required to add other custom agents?

Thank you!

**Comment by arabcoders** [comment, marked as answer]
Hi,

its not difficult task depending on how complex the GUID is, first we need mapping in

https://github.com/arabcoders/watchstate/blob/master/src/Libs/Guid.php

And parsing implementation in backends:

For plex
 
https://github.com/arabcoders/watchstate/blob/master/src/Backends/Plex/PlexGuid.php

Emby & Jellyfin share the same implementation

https://github.com/arabcoders/watchstate/blob/master/src/Backends/Jellyfin/JellyfinGuid.php

**Reply by ssy080** [reply]
Thank you for the reply. 

I hope I am right in also editing "https://github.com/arabcoders/watchstate/blob/master/src/Backends/Plex/PlexClient.php" to add supported agents. I've built and deployed and started the import command. System says waiting on 164 requests, so let's see if I've done it right. 

As it seems, it is not really an instantaneous sync across servers but a scheduled sync? How long does it take for one request to be completed?

Example GUID is KD99814/1/1. I ignored 1/1 as these are season/episode and just tried to match the number after the KD. I've put it as '/KD(\d+)/i'

edit2:
[2024-09-06T03:53:30+00:00] NOTICE: SYSTEM: Waiting on '164' requests.
[2024-09-06T04:03:35+00:00] NOTICE: SYSTEM: Completed waiting on '164' requests in '605s'.
[2024-09-06T04:03:35+00:00] NOTICE: MAPPER: No changes detected.

wrote log to import.txt. Log says "show has no valid/supported external ids. { (guids: None)," so I'm assuming regex did not match. 

edit3:
It appears that the agent that I am using has different guid scheme than ones provided. Instead of one string of numbers for a  given episode, this agent provides a string of numbers as a show id, followed by a season, then episode, seprated by a "/" as shown above. Would the program still work if I adjust the code to accept 99814/1/1 as guid?

**Reply by arabcoders** [reply]
> Thank you for the reply. 
> 
> 
> 
> I hope I am right in also editing "https://github.com/arabcoders/watchstate/blob/master/src/Backends/Plex/PlexClient.php" to add supported agents. I've built and deployed and started the import command. System says waiting on 164 requests, so let's see if I've done it right. 
> 
> 
> 
> As it seems, it is not really an instantaneous sync across servers but a scheduled sync? How long does it take for one request to be completed?
> 
> 
> 
> Example GUID is KD99814/1/1. I ignored 1/1 as these are season/episode and just tried to match the number after the KD. I've put it as '/KD(\d+)/i'
> 
> 
> 
> edit2:
> 
> [2024-09-06T03:53:30+00:00] NOTICE: SYSTEM: Waiting on '164' requests.
> 
> [2024-09-06T04:03:35+00:00] NOTICE: SYSTEM: Completed waiting on '164' requests in '605s'.
> 
> [2024-09-06T04:03:35+00:00] NOTICE: MAPPER: No changes detected.
> 
> 
> 
> wrote log to import.txt. Log says "show has no valid/supported external ids. { (guids: None)," so I'm assuming regex did not match. 
> 
> 
> 
> edit3:
> 
> It appears that the agent that I am using has different guid scheme than ones provided. Instead of one string of numbers for a  given episode, this agent provides a string of numbers as a show id, followed by a season, then episode, seprated by a "/" as shown above. Would the program still work if I adjust the code to accept 99814/1/1 as guid?


Hi, the requests are done in the same time, however depending on how fast the network+db and how big response it will take time, however once you enable webhooks after initial import it should be fast. 

On GUID please give real example of one so i can recommend an approach. 


if it's similar to the standard ones i.e. com.plex.agents.foo://series/season/episode we have legacy parser for it 

https://github.com/arabcoders/watchstate/blob/a14d524d4de3e4a035535d216ece3a3e0e5e4637/src/Backends/Plex/PlexGuid.php#L267

Also for testing you should always reset the database as we cache series calls.

**Reply by ssy080** [reply]
Yes, it is like the standard ones. This time, I've just added com.plexapp.agents.sjva_agent_ktv under GUID_LEGACY list, but still the same result. 

An example would be:
<Video ratingKey="469720" key="/library/metadata/469720" skipParent="1" parentRatingKey="387532" grandparentRatingKey="387531" guid="com.plexapp.agents.sjva_agent://KD99814/1/107?lang=ko" parentGuid="com.plexapp.agents.sjva_agent://KD99814/1?lang=ko" grandparentGuid="com.plexapp.agents.sjva_agent://KD99814?lang=ko" type="episode" ....

GUID being com.plexapp.agents.sjva_agent://KD99814/1/107?lang=ko. I

**Reply by arabcoders** [reply]
> Yes, it is like the standard ones. This time, I've just added com.plexapp.agents.sjva_agent_ktv under GUID_LEGACY list, but still the same result.
> 
> An example would be: <Video ratingKey="469720" key="/library/metadata/469720" skipParent="1" parentRatingKey="387532" grandparentRatingKey="387531" guid="com.plexapp.agents.sjva_agent://KD99814/1/107?lang=ko" parentGuid="com.plexapp.agents.sjva_agent://KD99814/1?lang=ko" grandparentGuid="com.plexapp.agents.sjva_agent://KD99814?lang=ko" type="episode" ....
> 
> GUID being com.plexapp.agents.sjva_agent://KD99814/1/107?lang=ko. I

If you have added it correctly, it should have been working, try resetting your watchstate database, after run `state:import -s plex --debug --logfile ~/config/plex.txt` it's gonna be quite big, however it should contains if parsing is failing if not, checking one item via webui should tell you if the GUID has been parsed correctly.

**Reply by ssy080** [reply]
[2024-09-06T16:10:27+00:00] DEBUG: Processing 'Plex: NUC1' response payload. { (library: [ (id: 7), (title: [종영] 1.드라마), (type: show), (totalRecords: 60856), (url: --) ]), (segment: [ (number: 53), (of: 61), (size: 1000) ]), (body: [ (ratingKey: 323371), (key: /library/metadata/323371), (skipParent: true), (parentRatingKey: 323364), (grandparentRatingKey: 323363), (guid: com.plexapp.agents.sjva_agent://KD93522/1/7?lang=ko), (parentGuid: com.plexapp.agents.sjva_agent://KD93522/1?lang=ko), (grandparentGuid: com.plexapp.agents.sjva_agent://KD93522?lang=ko), (type: episode), (title: 2022-07-28), (grandparentKey: /library/metadata/323363), (parentKey: /library/metadata/323364), (grandparentTitle: 나를 사랑하지 않는 X에게), (parentTitle: Season 1), (contentRating: 15세 관람가), (summary: 노트가 사라진 뒤, 시호의 응원을 받으며 그동안 자신이 정말 하고 싶었던 작사와 작곡에 집중하는 희수. 하지만 동아리 선배 도빈의 갑작스런 고백으로 시호의 오해를 산다.), (index: 7), (parentIndex: 1), (year: 2022), (thumb: /library/metadata/323371/thumb/1717744936), (art: /library/metadata/323364/art/1717744936), (parentThumb: /library/metadata/323364/thumb/1717744936), (grandparentThumb: /library/metadata/323363/thumb/1717744936), (grandparentArt: /library/metadata/323363/art/1717744936), ....

[2024-09-06T16:10:27+00:00] INFO: Ignoring 'Plex: NUC1' '나를 사랑하지 않는 X에게 - (01x007)'. No valid/supported external ids. { (guids: [ (0: com.plexapp.agents.sjva_agent://KD93522/1/7?lang=ko) ]), (library: [ (id: 7), (title: [종영] 1.드라마), (type: show), (totalRecords: 60856), (url: --) ]), (segment: [ (number: 53), (of: 61), (size: 1000) ]), (item: [ (id: 323371), (type: episode) ]) }

I've stopped the container, deleted db, then restarted the container. Did not run the import again but it seems like it still prints the logs. Is there another way to reset the database?

Above is the debug log. I am assuming it is not working as GUID after agent begins with letters KD.

**Reply by arabcoders** [reply]
Could you push your changes to github repo and link them here so review them? you or i might have missed something

**Reply by arabcoders** [reply]
I've seen the commits in your fork, if thats still the case and you have `com.plexapp.agents.sjva_agent_ktv` it's wrong and must be set to `com.plexapp.agents.sjva_agent` & this `'ktv' => Guid::GUID_KTV,` should be `'sjva_agent' => Guid::GUID_KTV,`

**Reply by ssy080** [reply]
It works! Thank you very much.

**Reply by ssy080** [reply]
Little off topic, but since I have to associate the backend with a user (and it only brings up users in the plex home group but not the ones with just shared access), does this mean that it only supports one user? Not all the users with access to the library?

Found partial answer on your discord: needs one WS instance per user. But still don't know if we can add users that are not in our home group.

**Reply by arabcoders** [reply]
No worries. You need their plex-token if they are not on your plex home. https://github.com/arabcoders/watchstate/blob/master/FAQ.md#how-do-i-migrate-invited-friends-ie-external-user-data-from-from-plex-to-embyjellyfin

**Reply by arabcoders** [reply]
@ssy080 this might be of interest to you #550
